src/tests/Microsoft.PowerFx.Interpreter.Tests/CustomFunctions.cs

```diff
@@ -44,6 +44,138 @@ public static StringValue Execute(NumberValue x, BooleanValue b)
+            config.AddFunction(new TestCallbackFunction());
+            var engine = new RecalcEngine(config);
+
+            // Shows up in enuemeration
+            var func = engine.GetAllFunctionNames().First(name => name == "TestCallback");
+            Assert.NotNull(func);
+
+            var result = engine.Eval("TestCallback(1=2)");
+            Assert.Equal(false, result.ToObject());
+        }
+
+        private class TestCallbackFunction : ReflectionFunction
+        {
+            // Must have "Execute" method. 
+            // Any arg can be a boolean callback function.
+            public static BooleanValue Execute(Func<Task<BooleanValue>> expression)
+            {
+                return expression().Result;
+            }
+        }
+
+        [Fact]
+        public async void CustomFunctionAsync_CallBack()
+        {
+            var config = new PowerFxConfig(null);
+            config.AddFunction(new WaitAsyncFunction());
+            config.AddFunction(new HelperFunction(x => FormulaValue.New(x.Value + 1)));
+            var engine = new RecalcEngine(config);
+
+            // Shows up in enuemeration
+            var func = engine.GetAllFunctionNames().First(name => name == "WaitAsync");
+            Assert.NotNull(func);
+
+            // Can be invoked. 
+            using var cts = new CancellationTokenSource();
+            var result = engine.EvalAsync("WaitAsync(Helper() = 3)", cts.Token);
+            Assert.Equal(true, (await result).ToObject());
+            Assert.True(result.IsCompletedSuccessfully);
+        }
+        
+        private class WaitAsyncFunction : ReflectionFunction
+        {
+            // Must have "Execute" method. 
+            // Cancellation Token must be the last argument for custom async function.
+            // Any arg can be a boolean callback function.
+            public static async Task<BooleanValue> Execute(Func<Task<BooleanValue>> expression, CancellationToken cancellationToken)
+            {
+                while (!(await expression()).Value) 
+                {
+                }
+
+                return FormulaValue.New(true);
+            }
+        }
+
+        private class HelperFunction : ReflectionFunction
+        {
+            private readonly Func<NumberValue, NumberValue> _func;
+            private NumberValue _counter;
+
+            public HelperFunction(Func<NumberValue, NumberValue> func)
+            {
+                _func = func;
+                _counter = FormulaValue.New(0);
+            }
+              
+            public NumberValue Execute()
+            {
+                _counter = _func(_counter);
+                return _counter;
+            }
+        }
+
+        [Fact]
+        public async void CustomFunctionAsync_CallBack_Invalid()
+        {
+            var config = new PowerFxConfig(null);
+            Assert.Throws<InvalidOperationException>(() => config.AddFunction(new InvalidTestCallbackAsyncFunction()));
```

Comment:
>InvalidOperationException [](http://example.com/codeflow?start=26&length=25)

You can check the error message here - this isn't getting caught by your new check (see comment about generics), but rather by an existing check.